src/DelayBucket.cc

```diff
@@ -25,29 +27,42 @@ DelayBucket::stats(StoreEntry *entry)const
 void
 DelayBucket::update(DelaySpec const &rate, int incr)
 {
-    if (rate.restore_bps != -1 &&
-            (level() += rate.restore_bps * incr) > rate.max_bytes)
-        level() = rate.max_bytes;
+    // make sure an undefined NaturalProduct() result (below) implies overflow
+    if (rate.restore_bps < 0)
+        return;
+
+    const auto oldLevel = level();
+    if (const auto delta = IncreaseProduct(rate.restore_bps, incr)) {
+        if (const auto newLevel = IncreaseSum(oldLevel, delta.value())) {
+            level() = newLevel.value();
```

Comment:
We should use our own `level_` to simplify further, I think:

```suggestion
    if (const auto delta = IncreaseProduct(rate.restore_bps, incr)) {
        if (const auto newLevel = IncreaseSum(level_, delta.value())) {
            level_ = newLevel.value();
```